extra/mediawiki/api/api.factor

```diff
@@ -0,0 +1,216 @@
+USING: arrays accessors assocs calendar combinators
+continuations formatting http http.client io json.reader kernel
+locals make math math.parser namespaces oauth1 prettyprint
+sequences strings system threads ;
+IN: mediawiki.api
+
+TUPLE: oauth-login consumer-token consumer-secret access-token
+access-secret ;
+TUPLE: password-login username password ;
+
+C: <oauth-login> oauth-login
+C: <password-login> password-login
+
+SYMBOLS: basetimestamp endpoint botflag contact cookies
+curtimestamp oauth-login password-login csrf-token ;
+
+t botflag set-global
+
+<PRIVATE
+
+: prepare ( params -- params' )
+    [ {
+        { [ dup t = ] [ drop "true" ] }
+        { [ dup number? ] [ number>string ] }
+        { [ dup string? ] [ ] }
+        { [ dup sequence? ] [
+            [ {
+                { [ dup number? ] [ number>string ] }
+                [ ]
+            } cond ] map "|" join
+        ] }
+    } cond ] assoc-map ;
+
+: <api-request> ( params -- request )
+        {
+            { "format" "json" }
+            { "formatversion" 2 }
+            { "maxlag" 5 }
+        } swap assoc-union prepare
+        endpoint get
+    <post-request>
+        contact get vm-version vm-git-id 7 head
+        "%s Factor/%s %s mediawiki.api" sprintf "User-Agent"
+        set-header ;
+
+: oauth-post ( params -- response data )
+    oauth-login get
+        dup consumer-token>>
+        over consumer-secret>> <token> consumer-token set
+        dup access-token>>
+        swap access-secret>> <token> access-token set
+    <api-request>
+        <oauth-request-params> set-oauth
+    http-request ;
+
+: cookie-post* ( params -- assoc )
+    <api-request>
+        cookies get >>cookies
+    http-request [ cookies>> cookies set-global ] dip json> ;
+
+: login-token ( -- token )
+    {
+        { "action" "query" }
+        { "meta" "tokens" }
+        { "type" "login" }
+    } cookie-post*
+    "query" "tokens" "logintoken" [ of ] tri@ ;
+
+: login ( -- cookies )
+    [
+        "login" "action" ,,
+        password-login get dup username>> "lgname" ,,
+        password>> "lgpassword" ,,
+        login-token "lgtoken" ,,
+    ] { } make cookie-post* drop cookies get ;
+
+: cookie-post ( params -- response data )
+    <api-request>
+        cookies get [ login ] unless* >>cookies
+    http-request ;
+
+: anon-post ( params -- response data )
+    <api-request> http-request ;
+
+: code-200? ( response assoc -- ? )
+    over code>> dup 200 = dup [ 3nip ] [
+        -roll "http status code %d" printf
+        swap header>> [ "=" glue print ] assoc-each
+        ...
+        10 minutes sleep
+    ] if ;
+
+: retry-after? ( response -- ? )
+    header>> "retry-after" of dup [ dup seconds sleep ] when ;
+
+: nonce-already-used? ( assoc -- ? )
+    "error" of
+    [ "code" of "mwoauth-invalid-authorization" = ]
+    [ "info" of "Nonce already used" swap subseq-start ] bi
+    and ;
+
+: readonly? ( assoc -- ? )
```

Comment:
Maybe the "is this readonly?" code could be separate from "sleep-if-readonly".

Reply:
Shouldn't we then also separate `code-200?` and `retry-after?`?

Reply:
If you want?  I was thinking that the sleep is part of a user control flow separate from request/response logic, but it's just thinking out loud.

Reply:
OK, I don't see how this could be meaningfully separated then.